chainer/functions/connection/multihead_attention.py

```diff
@@ -0,0 +1,298 @@
+    add_zero_attn=False,      # type: bool
+    attention_dropout=0,      # type: float
+    post_dropout=0,           # type: float
+    key_padding_mask=None,    # type: tp.Optional[InputType]
+    attn_mask=None,           # type: tp.Optional[InputType]
+    dot_product_scaler=None,  # type: tp.Optional[float]
+    softmax_scaler=1.0,       # type: float
+    return_weights=True       # type: bool
+):
+    # type: (...) -> tp.Tuple[variable.Variable, variable.Variable]  # NOQA
+    """Multi-head Attention forward function.
+
+    Args:
+        query (:class:`~chainer.Variable` or :ref:`ndarray`):
+            A batch of query vectors whose shape is :math:`(L, B, E)` where
+            :math:`L` is the target sequence length, :math:`B` is the
+            batch size, and :math:`E` is the embedding size.
+        key (:class:`~chainer.Variable` or :ref:`ndarray`)
+            A batch of key vectors whose shape is :math:`(S, B, E)` where
+            :math:`S` is the source sequence length, :math:`B` is the
+            batch size, and :math:`E` is the embedding size.
+        value (:class:`~chainer.Variable` or :ref:`ndarray`)
+            A batch of value vectors whose shape is :math:`(S, B, E)` where
+            :math:`S` is the source sequence length, :math:`B` is the
+            batch size, and :math:`E` is the embedding size.
+        expected_embedding_size (int): Total number of units of the model.
+        n_head (int): The number of parallel attention heads.
+        proj_in_W (:obj:`tuple`, :class:`~chainer.Variable` or :ref:`ndarray`):
+            Weight(s) to project `query`, `key`, and `value` vectors.
+            If the input sizes of `query`, `key`, and `value` are different,
+            this should be the tuple of three weights, otherwise, one weight.
+        proj_in_b (:class:`~chainer.Variable` or :ref:`ndarray`):
+            Bias added to projected `query`, `key`, and `value` vectors.
+        add_zero_attn (bool): If ``True``, add a new batch of zeros to
+            the key and value sequences at axis=1.
+        attention_dropout (float): Dropout ratio at the attention layer.
+        post_dropout (float): Dropout ratio at the output.
+        proj_out_W (:class:`~chainer.Variable` or :ref:`ndarray`):
+            Weight to project attention.
+        proj_out_b (:class:`~chainer.Variable` or :ref:`ndarray`):
+            Bias for projected attention.
+        key_padding_mask (:class:`~chainer.Variable` or :ref:`ndarray`):
+            If not ``None``, specified padding elements in the key
+            will be ignored by the attention.
+            The shape is :math:`(B, S)` where :math:`B` is the batch size,
+            and :math:`S` is the source sequence length.
+        attn_mask (:class:`~chainer.Variable` or :ref:`ndarray`):
+            Mask help attention ignores certain positions.
+            The shape is :math:`(L, L)` where :math:`L` is
+            the target sequence length.
```

Comment:
> This shape should be `(L, S)`?

Thanks.



> And, add an explanation that this value is "added" into attention values before softmax. Maybe, typical users expect the format of a mask for ignorance as multiplication. Adding a typical usage (e.g. -inf means setting ignoring the position) could be nice.

Yes, yes, yes!
I got confused why masking was done by addition instead of multiplication.